File: football_agents/edge_quality.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .closing_line_proxy import ClosingLineProxy
from .probability_uncertainty import ProbabilityUncertainty


@dataclass
class EdgeQualityOutput:
    outcome: str
    official_sp: float
    break_even_probability: float
    estimated_probability: float
    lower_bound_probability: float
    upper_bound_probability: float
    expected_ev: float
    lower_bound_ev: float
    upper_bound_ev: float
    expected_closing_edge: float | None
    clv_win_probability: float | None
    edge_quality_score: float
    edge_quality_level: str
    edge_noise_risk: str
    adaptive_threshold: float
    passes_true_odds_filter: bool
    reasons: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def calculate_edge_quality(outcome: str, official_sp: float, estimated_probability: dict[str, float], uncertainty: ProbabilityUncertainty, closing_proxy: ClosingLineProxy | None, context: dict[str, Any] | None, adaptive_threshold: float) -> EdgeQualityOutput:
    context = context or {}
    key = outcome.lower()
    p = estimated_probability[key]
    lower = uncertainty.lower[key]
    upper = uncertainty.upper[key]
    expected_ev = p * official_sp - 1
    lower_ev = lower * official_sp - 1
    upper_ev = upper * official_sp - 1
    score = 50.0
    reasons: list[str] = []
    warnings: list[str] = []
    if expected_ev > adaptive_threshold + 0.03:
        score += 15
    elif expected_ev > adaptive_threshold:
        score += 8
    if lower_ev > 0:
        score += 20
    else:
        score -= 25
        reasons.append("lowerBoundEV <= 0")
    clv_prob = closing_proxy.clv_win_probability if closing_proxy and closing_proxy.available else None
    closing_edge = closing_proxy.expected_closing_edge if closing_proxy and closing_proxy.available else None
    if clv_prob is not None and clv_prob > 0.55:
        score += 10
    if closing_edge is not None:
        score += 10 if closing_edge > 0 else -20
    score += 10 * float(context.get("method_agreement_score", 0))
    if str(context.get("external_market_quality") or "").upper() == "HIGH":
        score += 10
    if str(context.get("historical_bucket") or "").upper() == "POSITIVE":
        score += 10
    if str(context.get("model_disagreement") or "").upper() == "HIGH":
        score -= 20
        reasons.append("model disagreement high")
    if str(context.get("market_deviation") or "").upper() == "SUSPICIOUS":
        score -= 15
    if str(context.get("pure_model_reliability") or "").upper() == "LOW":
        score -= 15
    if str(context.get("lineup_risk") or "").upper() == "HIGH":
        score -= 10
    if str(context.get("fatigue_risk") or "").upper() == "HIGH":
        score -= 10
    if key == "draw" and not context.get("draw_calibrator_support"):
        score -= 8
    if official_sp > 5 or official_sp < 1.30:
        score -= 10
    if str(context.get("sample_size") or "").upper() == "LOW":
        score -= 15
    score = max(0, min(100, score))
    level = "HIGH" if score >= 75 else "MEDIUM" if score >= 55 else "LOW" if score >= 35 else "NO_EDGE"
    noise = "HIGH" if uncertainty.overall_uncertainty >= 0.75 or str(context.get("model_disagreement") or "").upper() == "HIGH" else "MEDIUM" if uncertainty.overall_uncertainty >= 0.5 else "LOW"
    if expected_ev <= adaptive_threshold:
        reasons.append("expected EV below adaptive threshold")
    if noise == "HIGH":
        reasons.append("edge noise risk high")
    passes = expected_ev > adaptive_threshold and lower_ev > 0 and level in {"HIGH", "MEDIUM"} and noise != "HIGH" and not context.get("critic_blocked", False)
    return EdgeQualityOutput(key.upper(), official_sp, 1 / official_sp, p, lower, upper, expected_ev, lower_ev, upper_ev, closing_edge, clv_prob, score, level, noise, adaptive_threshold, passes, list(dict.fromkeys(reasons)), warnings)

File: frontend/src/algorithm/edgeQuality.ts
import type {EdgeQualityOutput, ProbabilityUncertainty, RiskLevel, ThreeWayOdds, ThreeWayProbability} from "../types";

const keyOf=(outcome:"HOME"|"DRAW"|"AWAY")=>outcome==="HOME"?"home":outcome==="DRAW"?"draw":"away";
export function calculateEdgeQuality(outcome:"HOME"|"DRAW"|"AWAY",officialSp:number,estimated:ThreeWayProbability,uncertainty:ProbabilityUncertainty,context:{adaptiveThreshold:number;methodAgreementScore?:number;externalMarketQuality?:string;modelDisagreement?:RiskLevel;pureModelReliability?:RiskLevel;lineupRisk?:RiskLevel;fatigueRisk?:RiskLevel;sampleSize?:RiskLevel;drawCalibratorSupport?:boolean;criticBlocked?:boolean;expectedClosingEdge?:number|null;clvWinProbability?:number|null}):EdgeQualityOutput{
  const key=keyOf(outcome),p=estimated[key],lower=uncertainty.lower[key],upper=uncertainty.upper[key],expectedEv=p*officialSp-1,lowerBoundEv=lower*officialSp-1,upperBoundEv=upper*officialSp-1,reasons:string[]=[],warnings:string[]=[];
  let score=50;if(expectedEv>context.adaptiveThreshold+.03)score+=15;else if(expectedEv>context.adaptiveThreshold)score+=8;
  if(lowerBoundEv>0)score+=20;else{score-=25;reasons.push("lowerBoundEV <= 0")}
  if((context.clvWinProbability??.5)>.55)score+=10;
  if(context.expectedClosingEdge!==undefined&&context.expectedClosingEdge!==null)score+=context.expectedClosingEdge>0?10:-20;
  score+=10*(context.methodAgreementScore??0);
  if(context.externalMarketQuality==="HIGH")score+=10;
  if(context.modelDisagreement==="HIGH"){score-=20;reasons.push("模型分歧过高")}
  if(context.pureModelReliability==="LOW")score-=15;
  if(context.lineupRisk==="HIGH")score-=10;if(context.fatigueRisk==="HIGH")score-=10;
  if(outcome==="DRAW"&&!context.drawCalibratorSupport)score-=8;
  if(officialSp>5||officialSp<1.3)score-=10;
  if(context.sampleSize==="LOW")score-=15;
  score=Math.max(0,Math.min(100,score));
  const edgeQualityLevel=score>=75?"HIGH":score>=55?"MEDIUM":score>=35?"LOW":"NO_EDGE";
  const edgeNoiseRisk:RiskLevel=uncertainty.overallUncertainty>=.75||context.modelDisagreement==="HIGH"?"HIGH":uncertainty.overallUncertainty>=.5?"MEDIUM":"LOW";
  if(expectedEv<=context.adaptiveThreshold)reasons.push("expected EV 未超过 adaptive threshold");
  if(edgeNoiseRisk==="HIGH")reasons.push("edge 噪声风险过高");
  const passesTrueOddsFilter=expectedEv>context.adaptiveThreshold&&lowerBoundEv>0&&(edgeQualityLevel==="HIGH"||edgeQualityLevel==="MEDIUM")&&edgeNoiseRisk!=="HIGH"&&!context.criticBlocked;
  return{outcome,officialSp,breakEvenProbability:1/officialSp,estimatedProbability:p,lowerBoundProbability:lower,upperBoundProbability:upper,expectedEv,lowerBoundEv,upperBoundEv,expectedClosingEdge:context.expectedClosingEdge,clvWinProbability:context.clvWinProbability,edgeQualityScore:score,edgeQualityLevel,edgeNoiseRisk,adaptiveThreshold:context.adaptiveThreshold,passesTrueOddsFilter,reasons,warnings};
}
export const trueFairOdds=(p:ThreeWayProbability):ThreeWayOdds=>({home:1/Math.max(p.home,1e-12),draw:1/Math.max(p.draw,1e-12),away:1/Math.max(p.away,1e-12)});
